src/validation.jl

```diff
@@ -2,26 +2,36 @@
 
 export InvalidIRError
 
+function get_method_matches(@nospecialize(job::CompilerJob))
+    tt = typed_signature(job)
+
+    ms = Core.MethodMatch[]
+    for m in Base._methods_by_ftype(tt, -1, job.source.world)::Vector
+        m = m::Core.MethodMatch
+        push!(ms, m)
+    end
+
+    return ms
+end
+
+
 function check_method(@nospecialize(job::CompilerJob))
     isa(job.source.f, Core.Builtin) && throw(KernelError(job, "function is not a generic function"))
 
     # get the method
-    ms = Base.methods(job.source.f, job.source.tt)
+    ms = get_method_matches(job)
     isempty(ms)   && throw(KernelError(job, "no method found"))
     length(ms)!=1 && throw(KernelError(job, "no unique matching method"))
-    m = first(ms)
 
     # kernels can't return values
     if job.source.kernel
         cache = ci_cache(job)
         mt = method_table(job)
         interp = GPUInterpreter(cache, mt, job.source.world)
-        @static if v"1.8-beta2" <= VERSION < v"1.9-" || VERSION ≥ v"1.9.0-DEV.190"
-            # https://github.com/JuliaLang/julia/pull/44515
-            rt = Base.return_types(job.source.f, job.source.tt; interp)[1]
-        else
-            rt = Base.return_types(job.source.f, job.source.tt, interp)[1]
-        end
+        m = only(ms)
+        ty = Core.Compiler.typeinf_type(interp, m.method, m.spec_types, m.sparams)
```

Comment:
Same here, maybe it's better to use a local version of `return_types` that accepts function types and work on upstreaming that.